src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -566,24 +468,245 @@ object Genotype {
+           Gen.choose(0, m)));
+         dp <- Gen.option(Gen.choose(0, m));
+         gq <- Gen.option(Gen.choose(0, 10000));
+         pl <- Gen.oneOfGen(
+           Gen.option(Gen.buildableOfN[Array, Int](nGenotypes, Gen.choose(0, m))),
+           Gen.option(Gen.buildableOfN[Array, Int](nGenotypes, Gen.choose(0, 100))))) yield {
+      gt.foreach { gtx =>
+        pl.foreach { pla => pla(gtx) = 0 }
+      }
+      pl.foreach { pla =>
+        val m = pla.min
+        var i = 0
+        while (i < pla.length) {
+          pla(i) -= m
+          i += 1
+        }
+      }
+      val g = Genotype(gt, ad, dp.map(_ + ad.map(_.sum).getOrElse(0)), gq, pl)
+      g.check(nAlleles)
+      g
+    }
+  }
+
+  def genRealistic(nAlleles: Int): Gen[Genotype] = {
+    val nGenotypes = triangle(nAlleles)
+    for (callRate <- Gen.choose(0d, 1d);
+         alleleFrequencies <- Gen.buildableOfN[Array, Double](nAlleles, Gen.choose(1e-6, 1d)) // avoid divison by 0
+           .map { rawWeights =>
+           val sum = rawWeights.sum
+           rawWeights.map(_ / sum)
+         };
+         gt <- Gen.option(Gen.zip(Gen.chooseWithWeights(alleleFrequencies), Gen.chooseWithWeights(alleleFrequencies))
+           .map { case (gti, gtj) => Genotype.gtIndexWithSwap(gti, gtj) }, callRate);
+         ad <- Gen.option(Gen.buildableOfN[Array, Int](nAlleles,
+           Gen.choose(0, 50)));
+         dp <- Gen.choose(0, 30).map(d => ad.map(o => o.sum + d));
+         pl <- Gen.option(Gen.buildableOfN[Array, Int](nGenotypes, Gen.choose(0, 1000)).map { arr =>
+           gt match {
+             case Some(i) =>
+               arr(i) = 0
+               arr
+             case None =>
+               val min = arr.min
+               arr.map(_ - min)
+           }
+         });
+         gq <- Gen.choose(-30, 30).map(i => pl.map(pls => math.max(0, Genotype.gqFromPL(pls) + i)))
+    ) yield
+      Genotype(gt, ad, dp, gq, pl)
+  }
+
+  def genVariantGenotype: Gen[(Variant, Genotype)] =
+    for (v <- Variant.gen;
+         g <- Gen.oneOfGen(genExtreme(v.nAlleles), genRealistic(v.nAlleles), genDosage(v.nAlleles)))
+      yield (v, g)
+
+  def genArb: Gen[Genotype] =
+    for (v <- Variant.gen;
+         g <- Gen.oneOfGen(genExtreme(v.nAlleles), genRealistic(v.nAlleles), genDosage(v.nAlleles)))
+      yield g
+
+  implicit def arbGenotype = Arbitrary(genArb)
+}
+
+class GenericGenotype(private val _gt: Int,
+  private val _ad: Array[Int],
+  private val _dp: Int,
+  private val _gq: Int,
+  private val _px: Array[Int],
+  val fakeRef: Boolean,
+  val isDosage: Boolean) extends Genotype {
+
+  require(_gt >= -1, s"invalid _gt value: ${ _gt }")
+  require(_dp >= -1, s"invalid _dp value: ${ _dp }")
+
+  if (isDosage) {
+    require(_gq == -1)
+    if (_px == null)
+      require(_gt == -1)
+    else {
+      require(_px.sum == 32768)
+      require(_gt == Genotype.gtFromLinear(_px).getOrElse(-1))
+    }
+  }
+
+  override def equals(that: Any): Boolean = that match {
```

Comment:
Should a GenericGenotype and a MutableGenotype with the same per-field values compare equal?